docs/spec_draft.md

```diff
@@ -102,26 +102,192 @@ never produced as outputs.
+which the grid is split into process groups are shared between these ops and are
+described in this section. More formally, StableHLO supports the following
+four strategies.
+
+1) **cross_replica** (as in "only cross-replica communications will be happening
+within each process group"). This strategy takes `replica_groups` - a list of
+lists of replica ids - and computes a Cartesian product of `replica_groups` by
+`partition_ids`. `replica_groups` must have unique elements and cover all
+`replica_ids`. More formally:
+
+```Python
+def cross_replica(replica_groups: List[List[ReplicaId]]) -> List[List[ProcessId]]:
+  for replica_group in replica_groups:
+    for partition_id in partition_ids:
+      process_group = []
+      for replica_id in replica_group:
+        process_group.append((replica_id, partition_id))
+      yield process_group
+```
+
+For example, for `replica_groups = [[0, 1], [2, 3]]` and `num_partitions = 2`,
+`cross_replica` will produce
+`[[(0, 0), (1, 0)], [(0, 1), (1, 1)], [(2, 0), (3, 0)], [(2, 1), (3, 1)]]`.
+
+2) **cross_partition** (as in "only cross-partition communications will be
+happening within each process group"). This strategy takes `partition_groups` -
+a list of lists of partition ids - and computes a Cartesian product of
+`partition_groups` by `replica_ids`. `partition_groups` must have unique
+elements and cover all `partition_ids`. More formally:
+
+```Python
+def cross_partition(partition_groups: List[List[PartitionId]]) -> List[List[ProcessId]]:
+  for partition_group in partition_groups:
+    for replica_id in replica_ids:
+      process_group = []
+      for partition_id in partition_group:
+        process_group.append((replica_id, partition_id))
+      yield process_group
+```
+
+For example, for `partition_groups = [[0, 1]]` and `num_replicas = 4`,
+`cross_partition` will produce
+`[[(0, 0), (0, 1)], [(1, 0), (1, 1)], [(2, 0), (2, 1)], [(3, 0), (3, 1)]]`.
+
+3) **cross_replica_and_partition** (as in "both cross-replica and
+cross-partition communications will be happening within each process group").
+This strategy takes `replica_groups` - a list of lists of replica ids - and
+computes Cartesian products of each `replica_group` by `partition_ids`.
+`replica_groups` must have unique elements and cover all `replica_ids`.
+More formally:
+
+```Python
+def cross_replica_and_partition(replica_groups: List[List[ReplicaId]]) -> List[List[ProcessId]]:
+  for replica_group in replica_groups:
+    process_group = []
+    for partition_id in partition_ids:
+      for replica_id in replica_group:
+        process_group.append((replica_id, partition_id))
+    yield process_group
+```
+
+For example, for `replica_groups = [[0, 1], [2, 3]]` and `num_partitions = 2`,
+`cross_replica_and_partition` will produce
+`[[(0, 0), (1, 0), (0, 1), (1, 1)], [(2, 0), (3, 0), (2, 1), (3, 1)]]`.
+
+4) **flattened_ids**. This strategy takes `flattened_id_groups` - a list of
```

Comment:
Do we need to specify this (possibly)allows both cross-replica and cross-partition communication.

Reply:
Given the overall spirit of minimizing what we specify, I think we shouldn't call this out here.

By the way, your comment gave me motivation to review the descriptions of the previous three strategies, and I found a mistake. In "both cross-replica and cross-partition communications will be happening within each process group", "will" should really be "may".